### ISSUE TITLE
Tips of setting threshold

### ISSUE THREAD

**Issue by Jubi-Hwang** [zamanzadeh/CARLA#5]
Hello.
First, I would like to thank you for conducting such progressive and impressive research.
I was surprised when I saw the impact of point adjustment (PA) on an anomaly detection task.
I want to ask some tips of setting threshold in the situation that labels(anomaly detected or not) of the test dataset is unknown or private. Most of the anomaly detection research finds the threshold value by utilizing adjusted best F1 to evaluate the anomaly detection performance. However, this method can only be used when the labels are given. 
Now, I utilize train errors to set the threshold value like below code. Would it be useful to evaluate your model on a custom dataset in this way? I would like to ask for advice on how to set the threshold in this situation.

```
train_errors = []
with torch.no_grad():
    for batch in train_loader:
        inputs = batch["input"].cuda()
        targets = batch["target"].cuda() # original values
        outputs = model(inputs) # predicted values
        errors = torch.mean(torch.abs(targets - outputs), dim=1).cpu().numpy()
        train_errors.extend(errors)

threshold = np.mean(train_errors) + 2 * np.std(train_errors)
```

Thank you for reading my issue.

**Comment by zamanzadeh** [comment]
Thank you for your kind words about the research!

We use the best-F1 threshold (without adjustment) for our evaluation, a practical approach. This method allows for a fair comparison of anomaly detection performance by optimizing the threshold to maximize the F1 score during evaluation. However, we do not use any labels in the training phase, and all of the datasets except Yahoo that we used do not have labels in the training set. 

Setting a threshold when labels are unknown or private is indeed challenging. As shown in the code, this approach of using train errors to set the threshold aligns with our model’s approach and is reasonable. By calculating the mean and adding a multiple of the standard deviation (e.g., mean + 2 * std), you set a threshold that can capture anomalies based on the distribution of errors during training.

Here are some additional tips to refine your threshold setting:
- Validate on a Known Subset: If possible, use a small subset of data where labels are known to validate the threshold. This helps fine-tune the multiplier of the standard deviation (e.g., adjusting from 2 to 3) based on actual performance.
- Adjust for Sensitivity/Specificity Balance: Adjust the threshold accordingly depending on the criticality of false positives vs. false negatives in your application. Lowering the threshold increases sensitivity but also false positives, and vice versa.
- Use Percentiles: Instead of mean and standard deviation, consider using percentiles (e.g., 95th percentile) of the train errors to set the threshold, which can sometimes provide a more robust cutoff in skewed error distributions.
- Monitor Model Performance Over Time: Continuously monitor the model's performance and adjust the threshold as needed, especially if data distributions change over time.
